util/intern.h

```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+namespace atlas {
+namespace util {
+
+class StrRef {
+ public:
+  StrRef() = default;
+  bool operator==(const StrRef& rhs) const { return data == rhs.data; }
+  const char* data = nullptr;
+};
+
+const StrRef& intern_str(const char* string);
+const StrRef& intern_str(const std::string& string);
+const char* to_string(const StrRef& ref);
+}
+}
+
+namespace std {
+template <>
+struct hash<atlas::util::StrRef> {
+  size_t operator()(const atlas::util::StrRef& ref) const {
+    return reinterpret_cast<size_t>(ref.data);
```

Comment:
If I'm following this correctly, you are basically using the address as the hash value. I'm assuming that is for performance, but means that two `StrRef`s with the same data might not have the same hash value.